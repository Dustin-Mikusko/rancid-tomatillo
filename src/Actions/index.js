export const setMovies = movies => ({
  type: 'SET_MOVIES',
  movies
});

export const changeLoading = isLoading => ({
  type: 'CHANGE_LOADING',
  isLoading: !isLoading
});

export const addUser = user => ({
  type: 'ADD_USER',
  user
});
