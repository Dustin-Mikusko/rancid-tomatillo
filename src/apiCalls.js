export const getMovies = () => {
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/movies')
    .then(res => {
      if (!res.ok) {
        throw Error('Oops! The Box Office must be closed!')
      }
      return res.json();
    })
};

export const getUser = (user) => {
  const options = {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/login', options)
    .then(res => {
      if (!res.ok) {
        throw Error('Incorrect Email and/or Password')
      }
      return res.json();
    })
};

export const getUserRatings = id => {
  return fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${id}/ratings`).then(res => {
    if (!res.ok) {
      throw Error('Could not find your ratings')
    }
    return res.json();
  })
}
