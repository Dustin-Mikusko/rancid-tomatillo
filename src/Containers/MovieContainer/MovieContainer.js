import React, { Component } from 'react';
import { connect } from 'react-redux';
import Movie from '../../Components/Movie/Movie';
import Loading from '../../Components/Loading/Loading';
import { getMovies } from '../../apiCalls';
import { setMovies, changeLoading } from '../../Actions';

class MovieContainer extends Component {
  constructor() {
    super();
  }

  componentDidMount() {
    getMovies()
      .then(data => {
        this.props.setMovies(data.movies);
        this.props.changeLoading(this.props.isLoading)
      })
  }

  render() {
    if (this.props.isLoading) {
      return <Loading />
    }
    const displayMovies = this.props.movies.map( movie => {
      return (
        <Movie 
          id={movie.id}
          title={movie.title}
          poster={movie.poster_path}
          backdrop={movie.backdrop_path}
          releaseDate={movie.release_date}
          overview={movie.overview}
          avgRating={movie.avg_rating}
          key={movie.id}
        />
      )
    })

    return (
      <main>
        <h1>Oh hello - Movies Here</h1>
        {displayMovies}
      </main>
    )
  }
}

const mapStateToProps = state => ({
  movies: state.movies,
  isLoading: state.isLoading
});

const mapDispatchToProps = dispatch => ({
  changeLoading: (isLoading) => dispatch( changeLoading(isLoading)),
  setMovies: (movies) => dispatch(setMovies(movies))
});


export default connect(mapStateToProps, mapDispatchToProps)(MovieContainer);
