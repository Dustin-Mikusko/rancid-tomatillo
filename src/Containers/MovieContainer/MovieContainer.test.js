import React from 'react';
import { shallow } from 'enzyme';
import { MovieContainer, mapDispatchToProps } from './MovieContainer';
import { setMovies, changeLoading } from '../../Actions';
import { getMovies } from '../../apiCalls';

jest.mock('../../apiCalls.js')

describe('MovieContainer', () => {
  let wrapper, mockChangeLoading, mockSetMovies, mockMovies, mockIsLoading, mockUser;
  beforeEach( () => {
    mockChangeLoading = jest.fn();
    mockSetMovies = jest.fn();
    mockMovies = [{id: 1, title: "Movie Title", poster_path: "someURL", backdrop_path: "someURL", release_date: "2019-12-04", overview: "Some overview", average_rating: 6 }];
    mockIsLoading = false;
    mockUser = {
      id: 4,
      name: 'Greg',
      email: 'greg@turing.io',
      ratings: []
    }
    getMovies.mockImplementation( () => {
      return Promise.resolve({"movies": [{id: 1, title: "Movie Title", poster_path: "someURL", backdrop_path: "someURL", release_date: "2019-12-04", overview: "Some overview", average_rating: 6 }]})
    })
    wrapper = shallow(<MovieContainer 
      setMovies={mockSetMovies} 
      changeLoading={mockChangeLoading}
      movies={mockMovies}
      isLoading={mockIsLoading}
      user={mockUser}
      />)
  });

  it('should match the snapshot when content is loading', () => {
    wrapper = shallow(<MovieContainer isLoading={true}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot when isLoading is false', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot when there is a user and a user rating', () => {
    const mockUser = {
      id: 4,
      name: 'Greg',
      email: 'greg@turing.io',
      ratings: [{id: 1, user_id: 4, movie_id: 1, rating: 6, created_at: "someDate", updated_at: "someDate"}]
    }
    const wrapper = shallow(<MovieContainer user={mockUser} movies={mockMovies}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call getMovies when component mounts', () => {
    expect(getMovies).toHaveBeenCalled()
  });

  describe('mapDispatchToProps', () => {
    let mockDispatch;

    beforeEach( () => {
      mockDispatch = jest.fn();
    });

    it('should call dispatch with the changeLoading action when changeLoading is called', () => {
      const actionToDispatch = changeLoading(false);
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.changeLoading(false);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with the set movies action when setMovies is called', () => {
      const movies = [{title: 'The Marriage Story'}, {title: 'Knives Out'}, {title: 'It Chapter Two'}]
      const actionToDispatch = setMovies(movies)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.setMovies(movies);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

  });

});
