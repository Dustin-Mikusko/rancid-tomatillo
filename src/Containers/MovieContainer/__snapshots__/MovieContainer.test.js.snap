// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContainer should match the snapshot when content is loading 1`] = `ShallowWrapper {}`;

exports[`MovieContainer should match the snapshot when isLoading is false 1`] = `ShallowWrapper {}`;

exports[`MovieContainer should match the snapshot when there is a user and a user rating 1`] = `ShallowWrapper {}`;
